systeminit/si
# Restore validations

## Description

Validations:
- Re-introduce "check_validations" by executing the function immediately
  before code generation
- Add integration test for adding validations to a schema and then
  checking them for a given component
  - This test drove the "restore validations" goal of this commit
- Restore and refactor "check_validations" logic
  - Optimize core loop into one query
  - Use a cache for props already seen within the loop
- Move validations to their own func backend module
- Ensure ValidationError can be used for both kinds of validations

AWS:
- Add region field validation for the region schema
- Add integration test for region field validation
- Add "validationStringInStringArray" func, backend, etc.

Misc:
- Return the "created" bool for
  "FuncBinding::find_or_create_and_execute" (needed for validations
  restoration)
- Flesh out Tailscale doc
- Remove migration rust function (it was unneeded)

## Known Issue

The validations UI has likely not been used since the frontend refactor. Thus, while it is functional, it likely needs polish.

## Screenshots

Empty:

<img width="598" alt="Screen Shot 2022-09-26 at 6 14 56 PM" src="https://user-images.githubusercontent.com/39320683/192567962-24b41c20-3f65-4f52-979a-9cc6c4c0bb51.png">

Invalid:

<img width="780" alt="Screen Shot 2022-09-26 at 6 15 08 PM" src="https://user-images.githubusercontent.com/39320683/192567958-dba83e80-42c5-4712-baf3-521077aa3df8.png">

Valid:

<img width="519" alt="Screen Shot 2022-09-26 at 6 15 41 PM" src="https://user-images.githubusercontent.com/39320683/192567951-4057b651-6a01-4583-8e7d-9bc58f2156cf.png">

## Linear

Fixes ENG-539
Fixes ENG-540
Fixes ENG-558
